File: Python/merge_sort.py
sorting_list = [12, 13, 1, 4, 3, 8, 2]
final_list = [-1]* len(sorting_list)
def merge(list1, list2):
    # print(list1, list2)
    i = 0
    j = 0
    k = 0
    print(list1, list2, "::::", final_list)
    while i < len(list1) and j < len(list2):
        if list1[i] < list2[j]:
            final_list[k] = list1[i]
            i += 1
        else:
            final_list[k] = list2[j]
            j += 1
        k+=1
    while i < len(list1):
        final_list[k] = list1[i]
        i+=1
        k+=1
    while j < len(list2) :
        final_list[k] = list2[j]
        j += 1
        k += 1

def merge_sort(list):
    # print(left, right)
    if len(list)>1:
        mid = len(list) // 2
        Lhs = list[:mid]

        Rhs = list[mid:]
        merge_sort(Lhs)
        merge_sort(Rhs)
        list1 = Lhs
        list2 = Rhs
        # print(left, right, Lhs, Rhs)
        i = 0
        j = 0
        k = 0
        # print(list1, list2, "::::", final_list)
        while i < len(list1) and j < len(list2) :
            if list1[i] < list2[j] :
                list[k] = list1[i]
                i += 1
            else :
                list[k] = list2[j]
                j += 1
            k += 1
        while i < len(list1) :
            list[k] = list1[i]
            i += 1
            k += 1
        while j < len(list2) :
            list[k] = list2[j]
            j += 1
            k += 1
    # else:
    #     # print(left, right, list)
    #     return list


merge_sort(sorting_list)

print(sorting_list)

File: JAVA/merge_sort.java
import java.util.ArrayList;
import java.util.Arrays;

public class merge_sort {

    public static void merge (ArrayList<Integer> list1, Integer start1, Integer end1, Integer start2, Integer end2 ){

        ArrayList<Integer> temp = new ArrayList<>();

        Integer i = start1;
        Integer j = start2;

        while (i<= end1 && j <= end2){
            if (list1.get(i) < list1.get(j)){
                temp.add(list1.get(i));
                i++;
            }
            else if (list1.get(i) > list1.get(j)){
                temp.add(list1.get(j));
                j++;
            }
            else{
                temp.add(list1.get(i));
                temp.add(list1.get(j));
                i++; j++;
            }

        }

        while (i <= end1){
            temp.add(list1.get(i));
            i++;
        }
        while (j <= end2){
            temp.add(list1.get(j));
            j++;
        }

        Integer it_temp = 0;
        for(Integer k = start1; k <= end2; k++ ){
        list1.set(k, temp.get(it_temp
        ));
        it_temp++;}

    }

    public static void mergeSort(ArrayList<Integer> list1, Integer start, Integer end){
//        Integer start = 0;
//        Integer end = list1.size();

        if (start < end) {

            Integer mid;
            mid = (start + end) / 2;


            mergeSort(list1, start, mid);
            mergeSort(list1, mid + 1, end);
            merge(list1, start, mid, mid+1, end);
        }


    }

    public static void main (String[] args){
        ArrayList input_list = new ArrayList<Integer>(Arrays.asList(11,22,3,4,5,16,7,81,9));

        mergeSort(input_list, 0, input_list.size()-1);
        System.out.println(input_list);


    }
}
